### DISCUSSION TITLE
Packing a model with custom matcher

### DISCUSSION THREAD

**Discussion by maraimm** [explosion/spaCy#8913] [Help: Coding & Implementations] [answered]
Hi,

I am using Spacy3.1.1 to create a custom matcher that adds new entities.  I created a custom component and I register it to perform the matching in the main script. I then created a separated `my_component.py` to being able to package the component as a part of a new model. However, I am having two issues: 
1 -  The registered component that uses PhaseMatcher needs access to NLP model, when I initiate  the model the script does not work.
 
```
import spacy
from spacy.matcher import PhraseMatcher
from spacy.language import Language
# Import Span to slice the Doc
from spacy.tokens import Span

# Define the custom pipeline component
@Language.component("identify_books")
def identify_books33(doc):

	matcher = PhraseMatcher(nlp.vocab)
	# Apply the matcher to YOUR doc
	matches = matcher(doc)

	# List of book names to be matched 
	book_names = ['Pride and prejudice','Mansfield park','The Tale of Two cities','Great Expectations']

	# Creating pattern - list of docs through nlp.pipe() to save time
	book_patterns = list(nlp.pipe(book_names))
	
	# Adding the pattern to the matcher
	matcher.add("identify_books", None, *book_patterns)

	# Create a Span for each match and assign them under label "BOOKS"
	spans = [Span(doc, start, end, label="BOOKS") for match_id, start, end in matches]
	# Store the matched spans in doc.ents
	doc.ents = spans
	return doc
```

2. Packing the model.

    python3 -m spacy package MODEL_NAME compiled --code my_component.py --force

The model overwrites the Spacy model that I initialized it, for example `en_core_web_lg`. I tried to reset the model via uninstalling Spacy and reinstalling everything, but seems like the component is already registered. How can I start with a clean Spacy model?

    ValueError: [E002] Can't find factory for 'identify_books' for language English (en). This usually happens when spaCy calls `nlp.create_pipe` with a custom component name that's not registered on the current language class. If you're using a Transformer, make sure to install 'spacy-transformers'. If you're using a custom component, make sure you've added the decorator `@Language.component` (for function components) or `@Language.factory` (for class components).

Best,

**Comment by adrianeboyd** [comment, marked as answer]
It's going to be extremely inefficient to create a phrase matcher for each doc. If this is a realistic example, then you can use phrase patterns with the `entity_ruler` component without any custom components: https://spacy.io/usage/rule-based-matching#entityruler

If you do need a custom component:

* If you're only using `nlp.vocab`, then you can modify your `Language.component` to refer to `doc.vocab` instead, which should be identical.

* If you do need the whole `nlp` object to initialize, then you'd need to use `Language.factory` instead of `Language.component`. See the entity ruler for an example:

   https://github.com/explosion/spaCy/blob/a1e9f1946022112f7cd3c4c566691c5117a22850/spacy/pipeline/entityruler.py#L22-L53
